### ISSUE TITLE
Compiling with `CRASHPAD_ZLIB_SYSTEM=OFF`

### ISSUE THREAD

**Issue by daxpedda** [getsentry/sentry-native#354]
**Description**

Compiling without using the system zlib, but letting Crashpad build it, forces it to use SSE 4.1.
Preferably this should only be used on demand.

I assume that the problem is this:
https://github.com/getsentry/crashpad/blob/4ab7d237ff66ed82bc6e01ac159f1de81dae26cf/third_party/zlib/CMakeLists.txt#L47

What I find strange is that forcing SSE for the selected target CPU should be totally fine, but why doesn't it work with anything lower then SSE 4.1?

**When does the problem happen**

- [x] During build
- [ ] During run-time
- [ ] When capturing a hard crash

**Environment**

- OS: Linux
- Compiler: GCC 9
- CMake version and config: 3.18, SENTRY_BACKEND=crashpad

**Steps To Reproduce**

Just compile with `-DCRASHPAD_ZLIB_SYSTEM=OFF`.
Passing "-mpclmul" and "-msse4.1" CFLAGS fixes the issue. I tried other SSE versions without success.

**Log output**

```
2020-08-17T22:16:43.2113961Z [  6%] Building C object crashpad_build/third_party/zlib/CMakeFiles/crashpad_zlib.dir/zlib/crc_folding.c.o
2020-08-17T22:16:43.2114885Z crashpad_build/third_party/zlib/CMakeFiles/crashpad_zlib.dir/build.make:274: recipe for target 'crashpad_build/third_party/zlib/CMakeFiles/crashpad_zlib.dir/zlib/crc_folding.c.o' failed
2020-08-17T22:16:43.2115624Z CMakeFiles/Makefile2:550: recipe for target 'crashpad_build/third_party/zlib/CMakeFiles/crashpad_zlib.dir/all' failed
2020-08-17T22:16:43.2116186Z Makefile:146: recipe for target 'all' failed
2020-08-17T22:16:43.2116759Z 
2020-08-17T22:16:43.2117177Z --- stderr
2020-08-17T22:16:43.2117939Z In file included from /usr/lib/gcc/x86_64-linux-gnu/7/include/immintrin.h:39:0,
2020-08-17T22:16:43.2118631Z                  from /home/runner/work/sentry-contrib-native/sentry-contrib-native/sentry-contrib-native-sys/sentry-native/external/crashpad/third_party/zlib/zlib/crc_folding.c:23:
2020-08-17T22:16:43.2119342Z /home/runner/work/sentry-contrib-native/sentry-contrib-native/sentry-contrib-native-sys/sentry-native/external/crashpad/third_party/zlib/zlib/crc_folding.c: In function ‘fold_1’:
2020-08-17T22:16:43.2120228Z /usr/lib/gcc/x86_64-linux-gnu/7/include/wmmintrin.h:116:1: error: inlining failed in call to always_inline ‘_mm_clmulepi64_si128’: target specific option mismatch
2020-08-17T22:16:43.2120580Z  _mm_clmulepi64_si128 (__m128i __X, __m128i __Y, const int __I)
2020-08-17T22:16:43.2120848Z  ^~~~~~~~~~~~~~~~~~~~
2020-08-17T22:16:43.2121516Z /home/runner/work/sentry-contrib-native/sentry-contrib-native/sentry-contrib-native-sys/sentry-native/external/crashpad/third_party/zlib/zlib/crc_folding.c:71:17: note: called from here
2020-08-17T22:16:43.2122140Z      *xmm_crc3 = _mm_clmulepi64_si128(*xmm_crc3, xmm_fold4, 0x10);
2020-08-17T22:16:43.2122435Z                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-08-17T22:16:43.2123095Z In file included from /usr/lib/gcc/x86_64-linux-gnu/7/include/immintrin.h:39:0,
2020-08-17T22:16:43.2123856Z                  from /home/runner/work/sentry-contrib-native/sentry-contrib-native/sentry-contrib-native-sys/sentry-native/external/crashpad/third_party/zlib/zlib/crc_folding.c:23:
2020-08-17T22:16:43.2124559Z /usr/lib/gcc/x86_64-linux-gnu/7/include/wmmintrin.h:116:1: error: inlining failed in call to always_inline ‘_mm_clmulepi64_si128’: target specific option mismatch
2020-08-17T22:16:43.2124906Z  _mm_clmulepi64_si128 (__m128i __X, __m128i __Y, const int __I)
2020-08-17T22:16:43.2125167Z  ^~~~~~~~~~~~~~~~~~~~
2020-08-17T22:16:43.2125831Z /home/runner/work/sentry-contrib-native/sentry-contrib-native/sentry-contrib-native-sys/sentry-native/external/crashpad/third_party/zlib/zlib/crc_folding.c:70:17: note: called from here
2020-08-17T22:16:43.2126205Z      *xmm_crc0 = _mm_clmulepi64_si128(*xmm_crc0, xmm_fold4, 0x01);
2020-08-17T22:16:43.2126505Z                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-08-17T22:16:43.2127036Z In file included from /usr/lib/gcc/x86_64-linux-gnu/7/include/immintrin.h:39:0,
2020-08-17T22:16:43.2127723Z                  from /home/runner/work/sentry-contrib-native/sentry-contrib-native/sentry-contrib-native-sys/sentry-native/external/crashpad/third_party/zlib/zlib/crc_folding.c:23:
2020-08-17T22:16:43.2128441Z /usr/lib/gcc/x86_64-linux-gnu/7/include/wmmintrin.h:116:1: error: inlining failed in call to always_inline ‘_mm_clmulepi64_si128’: target specific option mismatch
2020-08-17T22:16:43.2128786Z  _mm_clmulepi64_si128 (__m128i __X, __m128i __Y, const int __I)
2020-08-17T22:16:43.2129037Z  ^~~~~~~~~~~~~~~~~~~~
2020-08-17T22:16:43.2129850Z /home/runner/work/sentry-contrib-native/sentry-contrib-native/sentry-contrib-native-sys/sentry-native/external/crashpad/third_party/zlib/zlib/crc_folding.c:71:17: note: called from here
2020-08-17T22:16:43.2130203Z      *xmm_crc3 = _mm_clmulepi64_si128(*xmm_crc3, xmm_fold4, 0x10);
2020-08-17T22:16:43.2130504Z                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-08-17T22:16:43.2131019Z In file included from /usr/lib/gcc/x86_64-linux-gnu/7/include/immintrin.h:39:0,
2020-08-17T22:16:43.2131909Z                  from /home/runner/work/sentry-contrib-native/sentry-contrib-native/sentry-contrib-native-sys/sentry-native/external/crashpad/third_party/zlib/zlib/crc_folding.c:23:
2020-08-17T22:16:43.2132625Z /usr/lib/gcc/x86_64-linux-gnu/7/include/wmmintrin.h:116:1: error: inlining failed in call to always_inline ‘_mm_clmulepi64_si128’: target specific option mismatch
2020-08-17T22:16:43.2132980Z  _mm_clmulepi64_si128 (__m128i __X, __m128i __Y, const int __I)
2020-08-17T22:16:43.2133264Z  ^~~~~~~~~~~~~~~~~~~~
2020-08-17T22:16:43.2134058Z /home/runner/work/sentry-contrib-native/sentry-contrib-native/sentry-contrib-native-sys/sentry-native/external/crashpad/third_party/zlib/zlib/crc_folding.c:70:17: note: called from here
2020-08-17T22:16:43.2134422Z      *xmm_crc0 = _mm_clmulepi64_si128(*xmm_crc0, xmm_fold4, 0x01);
2020-08-17T22:16:43.2134715Z                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-08-17T22:16:43.2135447Z make[2]: *** [crashpad_build/third_party/zlib/CMakeFiles/crashpad_zlib.dir/zlib/crc_folding.c.o] Error 1
2020-08-17T22:16:43.2135880Z make[1]: *** [crashpad_build/third_party/zlib/CMakeFiles/crashpad_zlib.dir/all] Error 2
2020-08-17T22:16:43.2136191Z make: *** [all] Error 2
```

**Comment by daxpedda** [comment]
I looked a bit around in the code, I don't know why CMake doesn't let me compile it with SSE2, according to the source code this is what it's made for.

**Comment by daxpedda** [comment]
Alright, this should be an easy fix, I got it all wrong.

So the version of zlib version of Google that is used here, uses a custom patch to enable SIMD support.
This is handled here (among others) in the original gn build file:
https://github.com/getsentry/crashpad/blob/4ab7d237ff66ed82bc6e01ac159f1de81dae26cf/third_party/zlib/BUILD.gn#L127-L145
```gn
  if (use_x86_x64_optimizations) {
    sources = [
      "crc_folding.c",
      "fill_window_sse.c",
    ]

    if (!is_win || is_clang) {
      cflags = [
        "-msse4.2",
        "-mpclmul",
      ]
    }
  }
```

But the CMake file from Sentry only says this:
https://github.com/getsentry/crashpad/blob/4ab7d237ff66ed82bc6e01ac159f1de81dae26cf/third_party/zlib/CMakeLists.txt#L44-L51
```cmake
    if(CMAKE_SYSTEM_PROCESSOR MATCHES "(x86_64)|(x86)|(i[3-7]86)|(AMD64)")
        target_sources(crashpad_zlib PRIVATE
            zlib/crc_folding.c
            zlib/fill_window_sse.c
            zlib/x86.c
            zlib/x86.h
        )
    endif()
```
Basically the C flags are missing.

For the issue of support, which was the main issue I was worried about, is that even though it needs SSE 4.2 to compile, it will still do the checks at runtime.
See the comment by the author of the patch: https://github.com/jtkukunas/zlib/issues/20#issuecomment-408976638
